packages/clarity-tutorials/contracts/rocket-tutorial/rocket-factory.scm

```diff
@@ -0,0 +1,117 @@
+;;  Copyright: (c) 2013-2019 by Blockstack PBC, a public benefit corporation.
+
+;;  This file is part of Blockstack.
+
+;;  Blockstack is free software. You may redistribute or modify
+;;  it under the terms of the GNU General Public License as published by
+;;  the Free Software Foundation, either version 3 of the License or
+;;  (at your option) any later version.
+
+;;  Blockstack is distributed in the hope that it will be useful,
+;;  but WITHOUT ANY WARRANTY, including without the implied warranty of
+;;  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+;;  GNU General Public License for more details.
+
+;;  You should have received a copy of the GNU General Public License
+;;  along with Blockstack. If not, see <http://www.gnu.org/licenses/>.
+
+;;;; Rocket-Factory
+
+;;; Storage
+(define-map orderbook
+  ((buyer principal))
+  ((rocket-id int) (ordered-at-block int) (ready-at-block int) (balance int) (size int)))
+(define-map auto-increments
+  ((id int))
+  ((value int)))
+
+;;; Constants
+(define funds-address 'SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR)
+
+;;; Internals
+
+;; Fetch, increment, update and return new rocket-id 
+;; returns: int
+(define (new-rocket-id)
+  (let ((rocket-id
+      (+ 1 (get value (fetch-entry auto-increments (tuple (id 0)))))))
+    (begin (set-entry! auto-increments
+      (tuple (id 0))
+      (tuple (value rocket-id)))
+    rocket-id)))
+
+;; Check if a given user can buy a new rocket
+;; args:
+;; @user (principal) the principal of the user
+;; returns: boolean
+(define (can-user-buy (user principal))
+  (let ((ordered-at-block
+      (get ordered-at-block 
+        (fetch-entry orderbook (tuple (buyer user))))))
+    (if (eq? ordered-at-block 'null) 'true 'false)))
+
+;; Check if a given user can claim a rocket previously ordered
+;; args:
+;; @user (principal) the principal of the user
+;; returns: boolean
+(define (can-user-claim (user principal))
+  (let ((ready-at-block
+      (get ready-at-block 
+        (fetch-entry orderbook (tuple (buyer user))))))
+    (>= block-height ready-at-block)))
+
+;; Order a rocket
+;; User not present in the orderbook have the ability to buy a new rocket.
+;; New rockets can be acquired using RKT tokens, price being a function of the size (size 5 = 5 RKT).
+;; Building a rocket takes time - it takes 10 blocks to build a rocket of size 10.
+;; Once the rocket is ready, users can get their rocket by calling the function "claim-rocket".
+;; args:
+;; @size (int) the size of the rocket (1 < size <= 20)
+;; returns: boolean
+(define-public (order-rocket (size int))
+  (let ((down-payment (/ size 2)))
+    (and 
```

Comment:
Please use `(if (immutable-conditions) (begin ((state-mutations) 'true) (...))` here and in `(claim-rocket)`.  Putting everything inside an `(and` makes this hard to read.

Reply:
I like the pattern:
```
// requirements
// mutations
```

However I think the current solution `(if (immutable-conditions) (begin ((state-mutations) 'true) (...))` should be improved - explained here: https://github.com/blockstack/blockstack-core/issues/972
Any take on that? @jcnelson @kantai

Reply:
My general programming philosophy is to minimize developer surprise and ambiguity -- both in Clarity scripts and elsewhere.  The pattern `IF [requirements] THEN [things to do if true] ELSE [things to do otherwise]` is much less surprising and much less ambiguous than a short-circuiting `[requirement 1] AND [requirement 2] AND [requirement 3] AND ... AND [1st thing to do if true] AND [2nd thing to do if true] AND [3rd thing to do if true] AND ...`.  In the former case, it's unambiguous which statements are `requirements` and which are `things to do`.  In the latter case, it's not only more ambiguous (since there's no language-level distinction), but also implicitly assumes that the circuit will be evaluated in a particular order and that the VM will short-circuit.

Reply:
If it was not clear, I definitely agree with you on the pattern `IF [requirements] THEN [things to do if true] ELSE [things to do otherwise]`.

My concern is about the behavior of `begin`. How would you safely handle:
```lisp
(begin
    (mutation-1)   ;; returns false
    (mutation-2)   ;; returns true
    'true)
```
With a `and(..)`?

Reply:
> My concern is about the behavior of begin. How would you safely handle:
```
(begin
    (mutation-1)   ;; returns false
    (mutation-2)   ;; returns true
    'true)
```

As the language is defined today, if `(mutation-1)` returns `'false`, and if you don't want `(mutation-2)` to execute unless `(mutation-1)` returns `'true`, you'd wrap `(mutation-2)` in an `(if`-statement that checks the result of `(mutation-1)` and only executes `(mutation-2)` if it is `'true`.

That said, doing this for every statement would be verbose and tedious.  In the future, we should add something like `(try!`, so you could do this:

```
(begin
   (try! (mutation-1))  ;; if (mutation-1) evals to 'false, then (try! causes the tx to abort
   (try! (mutation-2))
   'true)
```

Reply:
I moved your `try!` suggestion to https://github.com/blockstack/blockstack-core/issues/972 
Feedback addressed with https://github.com/blockstack/clarity-js-sdk/pull/4/commits/607d709c49b616f4c0bf81143555552c93255692

Reply:
This does not address my feedback.  I'm asking that you please not do `(and (mutation-1) (mutation-2) ...)`, since this isn't a programming pattern we want smart contract developers to emulate.  Please instead use something like `LET result-1 = mutation-1; IF !result-1 THEN abort; FI; LET result-2 = mutation-2; IF !result-2 THEN abort; FI; ...`.

Reply:
In a purely functional programming world -- using `(and ...)` to chain these kinds of things together is an appropriate pattern. `and` is always a short circuiting operation, and doing this avoids really deep `if` statement nesting. In a world with exception handling, you would use exception handling to similarly avoid the nested ifs.

Reply:
Even Rust has some (strange) handling of Result types for doing chained, dependent responses (see https://doc.rust-lang.org/rust-by-example/error/iter_result.html)